### PR TITLE
Redirect: support redirect stderr with piping stdout to next commands.

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description
Fixes: #10271

Given the following script:
```shell
# test.sh
echo aaaaa
echo bbbbb 1>&2
echo cc
```

This pr makes the following command possible:
```nushell
bash test.sh err> /dev/null | lines | each {|line| $line | str length}
```

## General idea behind the change:
When nushell redirect stderr message to external file
1. it take stdout of external stream, and pass this stream to next command, so it won't block next pipeline command from running.
2. relative stderr stream are handled by `save` command

These two streams are handled separately, so we need to delegate a thread to `save` command, or else we'll have a chance to hang nushell, we have meet a similar before: #5625.

### One case to consider
What if we're failed to save to an external stream? (Like we don't have a permission to save to a file)?
In this case nushell will just print a waning message, and don't stop the following scripts from running. 

# User-Facing Changes
## Before
```nushell
❯ bash test2.sh err> /dev/null | lines | each {|line| $line | str length}
aaaaa
cc
```

## After
```nushell
❯ bash test2.sh err> /dev/null | lines | each {|line| $line | str length}
╭───┬───╮
│ 0 │ 5 │
│ 1 │ 2 │
╰───┴───╯
```

BTY, after this pr, the following commands are impossible either, it's important to make sure that the implementation doesn't introduce too much costs:
```nushell
❯ echo a e> a.txt e> a.txt
Error:   × Can't make stderr redirection twice
   ╭─[entry #1:1:1]
 1 │ echo a e> a.txt e> a.txt
   ·                 ─┬
   ·                  ╰── try to remove one
   ╰────

❯ echo a o> a.txt o> a.txt
Error:   × Can't make stdout redirection twice
   ╭─[entry #2:1:1]
 1 │ echo a o> a.txt o> a.txt
   ·                 ─┬
   ·                  ╰── try to remove one
   ╰────
```

### CODE REVIEW

**Comment by 101313** [comment]
This has troubled me a lot, thanks man bless you

**Comment by WindSoilder** [comment]
I have no idea how to fix the CI....The CI runs failed because we save stderr in another thread, but we don't know if the thread is finished.  The machine is too busy to run this child thread

File: crates/nu-command/tests/commands/redirection.rs [hunk 1]
```diff
@@ -288,3 +288,38 @@ fn redirection_should_have_a_target() {
         );
     }
 }
+
+#[test]
+#[cfg(not(windows))]
```
Comment (sholderbach) [inline crates/nu-command/tests/commands/redirection.rs:293]: Would be great if we could convert this test to something that reliably runs on all platforms.

File: crates/nu-engine/src/eval.rs [hunk 2]
```diff
@@ -977,16 +1016,7 @@ pub fn eval_block(
...
-                || (elements_iter.peek().map_or(false, |next_element| {
-                    matches!(
-                        next_element,
-                        PipelineElement::Redirection(_, Redirection::Stdout, _)
-                            | PipelineElement::Redirection(_, Redirection::StdoutAndStderr, _)
-                            | PipelineElement::Expression(..)
-                            | PipelineElement::SeparateRedirection { .. }
-                    )
-                }));
+            let redirect_stdout = redirect_stdout || elements_iter.peek().is_some();
```
Comment (sholderbach) [inline crates/nu-engine/src/eval.rs:1019]: I assume the previous `matches!` is now handled somewhere else implicitly?

**Review by sholderbach — COMMENTED** [review]
Thanks for figuring this out and testing it!

I think we should move towards landing it soon :)

File: crates/nu-engine/src/eval.rs (hunk 2 above)
Comment (WindSoilder) [reply crates/nu-engine/src/eval.rs:1019]: Oh, wait!
Thanks for a good catch, I haven't consider it carefully...

Currently we need the redirection to make it work with this:
```nushell
cat a.txt err> /dev/null | lines
```
But it breaks this:
```nushell
def x [] {
    ^echo a.txt err> /dev/null   # it doesn't output a.txt anymore in this pr
    3
}
```
I need to look into this more

File: crates/nu-command/tests/commands/redirection.rs (hunk 1 above)
Comment (WindSoilder) [reply crates/nu-command/tests/commands/redirection.rs:293]: I'm thinking a way to make such output in our testbin.
I'd prefer to make it in separate pr, cause it requires a lot of changes in tests

File: crates/nu-engine/src/eval.rs (hunk 2 above)
Comment (WindSoilder) [reply crates/nu-engine/src/eval.rs]: It's a pity that the new implementation is more complicated :-(

File: crates/nu-parser/src/lite_parser.rs [hunk 3]
```diff
@@ -377,14 +388,28 @@ fn push_command_to(
...
-                    last_connector_span
-                        .expect("internal error: redirection missing span information"),
-                    Redirection::Stdout,
-                    command,
-                ));
+                let span = last_connector_span
+                    .expect("internal error: redirection missing span information");
+                if pipeline.exists(Redirection::Stdout) {
+                    return Some(ParseError::LabeledError(
+                        "Can't make stdout redirection twice".into(),
```
Comment (sophiajt) [inline crates/nu-parser/src/lite_parser.rs:395]: You can use positive text to make it more readable, like:

"Redirection can be set only once"

File: crates/nu-parser/src/lite_parser.rs [hunk 4]
```diff
@@ -377,14 +388,28 @@ fn push_command_to(
...
+                }
+
+                pipeline.push(LiteElement::Redirection(span, Redirection::Stdout, command));
             }
             TokenContents::ErrGreaterThan => {
+                let span = last_connector_span
+                    .expect("internal error: redirection missing span information");
+                if pipeline.exists(Redirection::Stderr) {
+                    return Some(ParseError::LabeledError(
+                        "Can't make stderr redirection twice".into(),
```
Comment (sophiajt) [inline crates/nu-parser/src/lite_parser.rs:408]: Same here

File: crates/nu-parser/src/lite_parser.rs (hunk 3 above)
Comment (WindSoilder) [reply crates/nu-parser/src/lite_parser.rs]: Thanks, it's a good improvement

**Pull request merged**